raiden_libs/network/matrix/client.py

```diff
@@ -21,31 +23,60 @@ class GMatrixHttpApi(MatrixHttpApi):
     A wrapper around MatrixHttpApi to limit the number
     of concurrent requests we make to the number of connections
     available to us in requests.Session connection pool size.
+
+    Args:
+        pool_maxsize: max size of underlying/session connection pool
+        retry_timeout: for how long should a single request be retried if it errors
+        retry_delay: delay to way between a retries inside a request.
+            If a callable, should return a float or an iterator of floats with each delay
     """
     def __init__(
             self,
             *args,
-            max_retries: int = 3,
             pool_maxsize: int = 10,
+            retry_timeout: int = 60,
+            retry_delay: Union[float, Callable[[], Union[float, Iterable[float]]]] = 1,
```

Comment:
Because, in raiden matrix, some calls to it are done before `start` is called, so we don't have access to the config parameters to be able to generate the exponential generator, and therefore (in this meantime) I need to return a constant value.

Reply:
I don't follow.

The signature `Callable[[], Iterable[float]]` should be sufficient to express all the use cases. For your example:

```python
from itertools import repeat

def constant_delay(delay):
   return lambda: repeat(delay)

GMatrixHttpApi(retry_delay=constant_delay(4))
```

Reply:
Do you really thing it's nicer than actually supporting returning a constant there?

Reply:
Yep, as a razor I would always go with the simpler type.

Your type represents three different options:

- float
- Callable[[], float]
- Callable[[], Iterable[float]]

This has a few problems. Three tests are necessary to cover the interface. The type is not easy to understand at a glance. It translates to an unnecessarily complex send method:

```
        if callable(self.retry_delay):
            retry_delay = self.retry_delay()  # should return a float or an iterable
        else:
            retry_delay = self.retry_delay  # is a float
        if not hasattr(retry_delay, '__iter__'):
            retry_delay = repeat(retry_delay)  # a float is repeated
```

Most of the lines in the code above is handling the special cases in the input, and have nothing to do with actually sending data.

`Callable[[], Iterable[float]]` is simpler, it's as expressive as `Union[float, Callable[[], Union[float, Iterable[float]]]]`. It's shorter, so it's user to understand. It doesn't have unions, so there are less inputs to test. And you can always add utility functions outside of the `_send` method that do the exactly the same, and is better separation of concerns.

Reply:
I don't agree with you in this case that simpler is better, as it means more complex elsewhere, but I don't have a strong feeling for this either, so I implemented it.